File: s2/c13.cpp
#include <unistd.h>
#include <cstdio>
#include <fstream>
#include <string>
#include "utils.h"
#include "aes128_ecb.h"
#include "aes128_cbc.h"
#include <sstream>
#include <map>
#include <stdexcept>

using namespace std;

class service {
	string m_key;
	aes128_ecb m_ecb;
	string encrypt(string input) {
		pad(input, 16);
		return m_ecb.encrypt(input);
	}
	string decrypt(const string& input) {
		string data = m_ecb.decrypt(input);
		data.erase(data.length()-data[data.length()-1]);
		return data;
	}
	map<string, string> splitUrlValues(const string& cookie) {
		map<string, string> ret;
		size_t pos = 0;
		size_t prev_pos = 0;
		while((pos = cookie.find_first_of("&=", prev_pos)) != string::npos) {
			string key = cookie.substr(prev_pos, pos-prev_pos);
			string value;
			prev_pos = pos + 1;
			if(cookie[pos] == '=') {
				if ((pos = cookie.find_first_of('&', prev_pos)) == string::npos) {
					pos = cookie.length();
				}
				value = cookie.substr(prev_pos, pos-prev_pos);
				prev_pos = pos + 1;
			}
			ret.insert(pair<string, string>(key, value));
		}
		if(prev_pos < cookie.length())
			ret.insert(pair<string, string>(cookie.substr(prev_pos, pos-prev_pos), string()));
		return ret;
	}
	public:
	service() : m_ecb(createRandomString(16)) {}
	string createProfile(const string& email) {
		ostringstream out;
		out << "email=" << url_encode(email);
		out << "&uid=" << rand()%1000000;
		out << "&role=user";
		return encrypt(out.str());
	}
	bool logIn(const string& cookie) {
		string data = decrypt(cookie);
		log("%s", data.c_str());
		map<string, string> userInfo = splitUrlValues(data);
		log("----------------");
		for(auto i = userInfo.begin(); i != userInfo.end(); i++) {
			log("%s:\t%s", i->first.c_str(), url_decode(i->second).c_str());
		}
		try {
			return userInfo.at("role") == "admin";
		} catch (out_of_range& e) {
		}
		return false;
	}
};

bool isECB(const string& cryptxt) {
	return (countRepetitions(cryptxt) > 1);
}

int main() {
	srand(time(NULL));
	service oracle;
	string test;
	string c1;
	test = "test@test.com";
	c1 = oracle.createProfile(test);
	oracle.logIn(c1);
	test = "test@test.com&role=admin";
	c1 = oracle.createProfile(test);
	oracle.logIn(c1);
	test = "";
	string cryptxt = oracle.createProfile(test);
	log("%zu", cryptxt.length());
	//	discover block size for oracle
	do {
		test.append(1, 'A');
		c1 = oracle.createProfile(test);
	} while(c1.length() == cryptxt.length());
	size_t blockLen = c1.length() - cryptxt.length();
	log("[>] Cipher block len: %zu", blockLen);
	//	check if ECB
	test = string(3*blockLen, 'A');
	if(!isECB(test)) {
		log("[-] Non ECB cipher");
		return 1;
	}
	log("[+] ECB mode discovered");
	//	No email validation so we can write anything there
	//	chunk "admin&uid=XXXXXX"
	test.clear();
	test.append(16-string("email=").length(), 'a');
	test.append("admin");
	c1 = oracle.createProfile(test).substr(16, 16);
	//	get valid padding
	test.append(7, 'A');
	string p1;
	do {
		p1 = oracle.createProfile(test).substr(48);
	} while(p1.length() < 16);
	c1.append(p1);
	//	chunk "uid=XXXXXX&role="
	test.clear();
	test.append(16-string("email=&").length(), 'a');
	string c2;
	do {
		c2 = oracle.createProfile(test).substr(0, 32);
		c2.append(c1);
	} while(!oracle.logIn(c2));
	return 0;
}

File: s2/c13.py
#!/usr/bin/python3

import py_common.cryptoUtils as cu
import base64
import string
import sys
from random import SystemRandom as SRand
import urllib.parse as url

class oracle:
    def __init__(self):
        self.key = cu.createRandomString(16)

    def createProfile(self, email):
        plain = cu.pad(self.profile_for(email).encode('ascii'), 16)
        return cu.encryptAES_ECB(plain, self.key)

    def logIn(self, data):
        plain = cu.decryptAES_ECB(data, self.key)
        print(plain)
        plain = plain[:-plain[-1]].decode('ascii')
        user = self.getUserInfo(plain)
        print("----------------\nUser:",user['role'],"\nUid:",user['uid'])
        print("email:",user['email'])
        return user['role'] == 'admin'

    def getUserInfo(self, data):
        user = dict()
        for i in data.split('&'):
            j = i.split('=')
            if len(j) != 2:
                continue
            user[j[0]] = url.unquote(j[1])
        return user

    def profile_for(self, email):
        email = url.quote(email)
        uid = SRand().randint(0, 1000000)
        role = 'user'
        return "email={}&uid={}&role={}".format(email, uid, role)

if __name__ == "__main__":
    o = oracle()
    cookie = o.createProfile("test@test.com")
    o.logIn(cookie)
    cookie = o.createProfile("test@test.com&role=admin")
    o.logIn(cookie)
    cryptxt = o.createProfile("")
    test = bytearray()
    while True:
        c1 = o.createProfile(bytes(test))
        if len(cryptxt) < len(c1):
            break
        test.append(ord('A'))
    blockLen = len(c1) - len(cryptxt)
    print("[>] Block len:",blockLen)
    test = b"A"*3*blockLen
    test += b"@gmail.com"
    if not cu.isECB(o.createProfile(test)):
        print("[-] Non ECB oracle")
        sys.exit(1)
    print("[+] ECB mode discovered")
    #   generate chunk: "admin&uid=YYYYYY&role", where YYYYYY is in <0, 10^7>, most likely 6 digits
    #                   "email=aaaaaaaaa@"
    #                   "aaaaa.admin&uid="
    #                   "YYYYYY&role=user"
    #                   "PPPPPPPPPPPPPPPP"
    #   @ gets urlencoded to %40
    e1 = "a" * (blockLen-len("email=")-3) + "@" + "admin"
    c1 = o.createProfile(e1)[16:32]
    #   generate padding chunk(\x10 * 16)
    e1 = "a" * (blockLen-len("email=")-3) + "@" + "aaaaa.admin"
    while True:
        #   ensure that padding chunk is what we want it to be
        p1 = o.createProfile(e1)[48:]
        if len(p1) == 16:
            break
    #   generate chunk: "uid=XXXXXX&role="
    #   we can't read id, so need to guess, but probability is with us
    #   few tries might be needed
    e1 = "b" * (blockLen - len("email=") - 3)  + "@" + "A"*11 + ".com"
    while True:
        c2 = o.createProfile(e1)[:48]
        fake = c2 + c1 + p1
        if o.logIn(fake):
            break
